src/internet_identity/src/state/flow_states.rs

```diff
@@ -0,0 +1,131 @@
+    /// Registration flows are removed on three conditions:
+    /// * The registration flow is completed successfully
+    /// * Once the flow expires
+    /// * The Internet Identity canister is upgraded
+    ///
+    /// The total number of flow states is limited by the registration rate limit.
+    registration_flows: HashMap<Principal, RegistrationFlowState>,
+
+    /// Deque to efficiently prune expired registration flows
+    ///
+    /// The total number of flow state expirations is limited by the registration rate limit.
+    expirations: VecDeque<FlowExpiration>,
+}
+
+/// State of a registration flow. The state expresses which action is expected next.
+#[derive(Debug, Clone)]
+pub enum RegistrationFlowState {
+    /// Client is expected to supply the `captcha_solution` using
+    /// `check_captcha`.
+    CheckCaptcha {
+        flow_created_timestamp_ns: u64,
+        captcha_solution: String,
+        captcha_png_base64: Base64,
+    },
+    /// Client is expected to finish the registration (supply an [authn_method](internet_identity_interface::internet_identity::types::AuthnMethodData))
+    /// using `identity_registration_finish`.
+    FinishRegistration { flow_created_timestamp_ns: u64 },
+}
+
+impl RegistrationFlowState {
+    pub fn create_at_timestamp_ns(&self) -> u64 {
+        match self {
+            RegistrationFlowState::CheckCaptcha {
+                flow_created_timestamp_ns,
+                ..
+            }
+            | RegistrationFlowState::FinishRegistration {
+                flow_created_timestamp_ns,
+            } => *flow_created_timestamp_ns,
+        }
+    }
+}
+
+impl FlowStates {
+    pub fn new_registration_flow(
+        &mut self,
+        principal: Principal,
+        flow_state: RegistrationFlowState,
+    ) -> Result<(), String> {
+        self.prune_expired_flows();
+        if self.registration_flows.contains_key(&principal) {
+            return Err("already exists".to_string());
+        }
+
+        self.expirations.push_back(FlowExpiration {
+            principal,
+            expiration: flow_state.create_at_timestamp_ns() + FLOW_EXPIRATION_NS,
+        });
+
+        self.registration_flows.insert(principal, flow_state);
+        Ok(())
+    }
+
+    pub fn update_registration_flow(
+        &mut self,
+        principal: Principal,
+        flow_state: RegistrationFlowState,
+    ) -> Result<(), String> {
+        self.prune_expired_flows();
+        if self.registration_flows.remove(&principal).is_none() {
+            return Err("no flow_state to update".to_string());
+        }
+        self.registration_flows.insert(principal, flow_state);
+        Ok(())
+    }
+
+    pub fn remove_registration_flow(&mut self, principal: &Principal) {
+        self.prune_expired_flows();
+        self.registration_flows.remove(principal);
+    }
+
+    pub fn registration_flow_state(&self, principal: &Principal) -> Option<RegistrationFlowState> {
```

Comment:
Should we also prune here?

Reply:
I don't think it is necessary, as we explicitly check `expiration` before returning anything.
IMHO it is more natural to prune alongside other modifications, rather than on read.